src/Framework/ITaskItemExtensions.cs

```diff
@@ -39,26 +37,9 @@ internal static class ITaskItemExtensions
             var list = new KeyValuePair<string, string>[customMetadata.Count];
             int i = 0;
 
-            foreach (string metadataName in customMetadata.Keys)
+            foreach (KeyValuePair<string, string> metadatum in customMetadata)
             {
-                string valueOrError;
-
-                try
-                {
-                    valueOrError = taskItem.GetMetadata(metadataName);
-                }
-                // Temporarily try catch all to mitigate frequent NullReferenceExceptions in
-                // the logging code until CopyOnWritePropertyDictionary is replaced with
-                // ImmutableDictionary. Calling into Debug.Fail to crash the process in case
-                // the exception occurres in Debug builds.
-                catch (Exception e)
-                {
-                    valueOrError = e.Message;
-                    Debug.Fail(e.ToString());
-                }
-
-                list[i] = new KeyValuePair<string, string>(metadataName, valueOrError);
-                i += 1;
+                list[i++] = new KeyValuePair<string, string>(metadatum.Key, metadatum.Value);
             }
```

Comment:
Can this not be simply
```C#
customMetadata.Select(metadatum => new KeyValuePair<string, string>(metadatum.Key, metadatum.Value)).ToArray();
```
?

Reply:
I don't think so, but I'm honestly not 100% sure why not. I'll admit it looks kinda stupid. An IDictionary is an IEnumerable that contains a collection of key-value pairs, per its documentation, but it doesn't have those as its explicit types, so it's an IEnumerable but not an IEnumerable<...>. Since it's an IEnumerable, you can loop over it, but Linq only works on IEnumerable<...>, I guess? Either way, I don't think this code really executes in practice, so although it'd be nice if it could be cleaner, I'm not too worried about it.

Reply:
Of course, you're right. I missed that it's the non-generic `IDictionary`. In that case, though, your new code would not work either. The `foreach` will fail to cast `DictionaryEntry` to `KeyValuePair<string, string>`.

Reply:
I was thinking we're ok because all the implementations of CopyCustomMetadata outside tests and Deprecated return IDictionarys that also implement IDictionary<string, string>. That made me think that as long as this compiled, it was ok. On the other hand, someone could theoretically implement their own custom CopyCustomMetadata and return, say, a Hashtable, and then it might crash. I'll change this to work for that case. Thanks for taking a look!

Reply:
Implementations that also support `IDictionary<string, string>` are caught by the `is IEnumerable<KeyValuePair<string, string>>` check above, so this code really deals with types that support only the non-generic `IDictionary`. Not sure how often it happens but it's safer to keep it in, for sure.